### PR TITLE
Nvfuser - Type Promotion Fix

### PR DESCRIPTION
**Pull request by rdspring1**
Fix Type Promotion failures in [Issue 76046](https://github.com/pytorch/pytorch/issues/76046)

1. Updated nvfuser type promotion rule for codegen kernel;
2. Updated casting for output of nvfuser kernel to respect profiling/TorchScript scalar type;
3. Updated type_inference.cpp to only update device/scalar_type when profiling information is missing.

Additional Type Promotion Fixes:
-  test_nvfuser_correctness_softmax_with_dtype_cuda_float32
-  test_nvfuser_correctness_softmax_with_dtype_cuda_bfloat16
-  test_nvfuser_correctness_softmax_with_dtype_cuda_float16
-  test_nvfuser_correctness_softmax_with_dtype_cuda_float32
-  test_nvfuser_correctness_log_softmax_dtype_cuda_bfloat16
-  test_nvfuser_correctness_log_softmax_dtype_cuda_bool
-  test_nvfuser_correctness_log_softmax_dtype_cuda_float16
-  test_nvfuser_correctness_log_softmax_dtype_cuda_float32
-  test_nvfuser_correctness_sum_cuda_int32
-  test_nvfuser_correctness_sum_to_size_cuda_int32

### CODE REVIEW

**Comment by mruberry** [comment]
Lintrunner failure is real. We have a utility to help run it locally now: https://github.com/pytorch/pytorch/wiki/lintrunner

**Comment by mruberry** [comment]
@jjsjann123 -- ping me when you're happy with this?

File: torch/csrc/jit/codegen/cuda/arith.cpp [hunk 1]
```diff
@@ -980,7 +1004,10 @@ Val* sub_alpha(Val* v1, Val* v2, Val* s) {
       "Alpha value should be a Scalar Valtype and not ",
       s->getValType().value());
 
-  auto vals = maybeBroadcast({v1, v2, s});
+  std::vector<Val*> operands = {v1, v2};
+  auto common_dtype = computeTypes(TypePromotion::default_op_config, operands);
+  auto casted_values = promoteValues({v1, v2, s}, common_dtype);
```
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/arith.cpp:1009]: DITTO

File: torch/csrc/jit/codegen/cuda/arith.cpp [hunk 2]
```diff
@@ -1028,7 +1055,10 @@ Val* addcmul(Val* v1, Val* v2, Val* v3, Val* s) {
       "Alpha value should be a Scalar Valtype and not ",
       s->getValType().value());
 
-  auto vals = maybeBroadcast({v1, v2, v3, s});
+  std::vector<Val*> operands = {v1, v2, v3};
+  auto common_dtype = computeTypes(TypePromotion::default_op_config, operands);
+  auto casted_values = promoteValues({v1, v2, v3, s}, common_dtype);
```
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/arith.cpp:1060]: DITTO

File: torch/csrc/jit/codegen/cuda/parser.cpp [hunk 3]
```diff
@@ -1977,12 +1959,24 @@ class IrParser {
...
 
+              auto data_type = DataType::Null;
+              if (const auto opt_ivalue = toIValue(node->input(2))) {
+                if (!opt_ivalue.value().isNone()) {
+                  data_type = aten_to_data_type(opt_ivalue->toScalarType());
+                }
+              }
+
+              input = (data_type != DataType::Null)
+                  ? optionalCastStrict(data_type, input)->as<TensorView>()
```
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/parser.cpp:1970]: Note: updated math type... we could stick with our math type in parser.

File: torch/csrc/jit/codegen/cuda/arith.cpp [hunk 4]
```diff
@@ -186,6 +186,58 @@ Val* newValLike(Val* val, DataType dtype) {
   return newScalar(vtype, dtype);
 }
 
+Val* getMinimumValue(DataType v) {
```
Comment (mruberry) [inline torch/csrc/jit/codegen/cuda/arith.cpp:189]: Not a requirement, but a brief comment for each function would be nice for readability/reviewability

File: torch/csrc/jit/codegen/cuda/arith.cpp [hunk 5]
```diff
@@ -696,88 +749,56 @@ TensorView* reductionOp(
...
+  if (dtype == DataType::Null) {
+    auto initial_v1_dtype = v1->getDataType().value();
+    if (isBooleanType(initial_v1_dtype) || isIntegralType(initial_v1_dtype)) {
+      dtype = DataType::Int;
+    }
+  }
+
+  // Cast input tensor to dtype before the operation is performed
+  if (dtype != DataType::Null) {
+    v1 = optionalCastStrict(dtype, v1)->as<TensorView>();
```
[the remark this answers is not in the document]
Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/arith.cpp:764]: We need to have better support for dtype argument in integration. e.g. currently some parse entry (i.e. sum) only fuses node when dtype are unspecified. :stuck_out_tongue:

**Review by jjsjann123 — APPROVED** [review]
LGTM.

**Review by mruberry — APPROVED** [review]
Stamped! Nice OpInfo skip fixes.

**Comment by mruberry** [comment]
@rdspring1 @jjsjann123 I think you can merge these PRs yourselves now using @pytochbot merge this please when you're ready. It just needs an approving review from someone with commit writes

**Pull request closed without merging**